### PR TITLE
Slasher: Refactor and add tests

### PR DESCRIPTION
**Pull request by nalepae**
**Please read this PR commit by commit.**

**What type of PR is this?**
Other

**What does this PR do? Why is it needed?**
This PR does not add any functional change, but:
- Add some logs
- Refactor some parts of the slasher (flatten happy path, variable renaming, etc...)
- Fix some tests

The last commit adds new tests, with some commented failing tests.
For each failing test, a ticket issue is linked.

A next PR will fix these tickets:
- https://github.com/prysmaticlabs/prysm/issues/13430
- https://github.com/prysmaticlabs/prysm/issues/13590
- https://github.com/prysmaticlabs/prysm/issues/13591
- https://github.com/prysmaticlabs/prysm/issues/13592

### CODE REVIEW

File: beacon-chain/slasher/chunks.go [hunk 1]
```diff
@@ -492,12 +516,12 @@ func (m *MinSpanChunksSlice) NextChunkStartEpoch(startEpoch primitives.Epoch) pr
...
 // If C = chunkSize is 3 epochs per chunk, and we input start epoch of chunk 1 which is 3. The next start
-// epoch is the start epoch of chunk 2, which is epoch 4. This is computed as:
+// epoch is the start epoch of chunk 2, which is epoch 6. This is computed as:
 //
 //	first_epoch(chunkIndex(startEpoch)+1)
 //	first_epoch(chunkIndex(3)+1)
 //	first_epoch(1 + 1)
-//	first_epoch(2)
+//	first_epoch(6)
 //	4
```
Comment (prestonvanloon) [inline beacon-chain/slasher/chunks.go:525]:
```suggestion
//	first_epoch(2)
//	6
```
I think this is what the author meant. The last value is the epoch. I read this as `first_epoch(chunk_2) => 6`

File: beacon-chain/slasher/receive.go [hunk 2]
```diff
@@ -92,55 +92,76 @@ func (s *Service) processQueuedAttestations(ctx context.Context, slotTicker <-ch
...
 			attestations := s.attsQueue.dequeue()
-			currentEpoch := slots.ToEpoch(currentSlot)
-			// We take all the attestations in the queue and filter out
-			// those which are valid now and valid in the future.
-			validAtts, validInFuture, numDropped := s.filterAttestations(attestations, currentEpoch)
 
-			deferredAttestationsTotal.Add(float64(len(validInFuture)))
-			droppedAttestationsTotal.Add(float64(numDropped))
+			// Process the retrieved attestations.
+			s.processAttestations(ctx, attestations, currentSlot)
```
Comment (prestonvanloon) [inline beacon-chain/slasher/receive.go:99]: Excellent refactoring here. This select case was a mess!

File: beacon-chain/slasher/chunks.go [hunk 3]
```diff
@@ -252,29 +263,42 @@ func (m *MaxSpanChunksSlice) CheckSlashable(
...
-			}, nil
-		}
+	if existingAttRecord == nil {
+		// This case should normally not happen. If this happen, it means we previously
+		// recorded in our min/max DB an distance corresponding to an attestaiton, but WITHOUT
+		// recording the attestation itself. As a consequence, we say there is no surrounded vote,
+		// but we log an error.
+		log.Errorf("No existing attestation record found for validator %d at target %d, while a surrounded vote was detected.",
+			validatorIdx, maxTarget,
+		)
```
Comment (prestonvanloon) [inline beacon-chain/slasher/chunks.go:284]: Avoid templated log messages, it makes it harder to monitor / aggregate these.

Prefer using logrus fields https://pkg.go.dev/github.com/sirupsen/logrus#Logger.WithFields

File: beacon-chain/slasher/detect_attestations.go [hunk 4]
```diff
@@ -27,51 +27,55 @@ func (s *Service) checkSlashableAttestations(
...
 
 	// Surrounding / surrounded votes
-	groupedAtts := s.groupByValidatorChunkIndex(atts)
-	log.WithField("numBatches", len(groupedAtts)).Debug("Batching attestations by validator chunk index")
-	start = time.Now()
-	batchTimes := make([]time.Duration, 0, len(groupedAtts))
+	groupedByValidatorChunkIndexAtts := s.groupByValidatorChunkIndex(atts)
+	log.WithField("numBatches", len(groupedByValidatorChunkIndexAtts)).Debug("Batching attestations by validator chunk index")
+	groupsCount := len(groupedByValidatorChunkIndexAtts)
+	batchDurations := make([]time.Duration, 0, len(groupedByValidatorChunkIndexAtts))
```
Comment (prestonvanloon) [inline beacon-chain/slasher/detect_attestations.go:39]: batchDurations is only used to calculate the totalBatchDuration. Why not just accumulate / update totalBatchDuration and skip this slice allocation?

File: beacon-chain/slasher/chunks.go (hunk 1 above)
Comment (nalepae) [reply beacon-chain/slasher/chunks.go:525]: I don't know how I ended up with this conclusion 😂

File: beacon-chain/slasher/chunks.go (hunk 3 above)
Comment (nalepae) [reply beacon-chain/slasher/chunks.go]: Fixed in this commit: https://github.com/prysmaticlabs/prysm/pull/13589/commits/bd02048eec6dace73b384f77ae1a31ce14070a61

File: beacon-chain/slasher/detect_attestations.go (hunk 4 above)
Comment (nalepae) [reply beacon-chain/slasher/detect_attestations.go]: Fixed in this commit: https://github.com/prysmaticlabs/prysm/pull/13589/commits/7acbc4d8a2465ee369a442f532e591f584dca297

**Review by rauljordan — APPROVED** [review]
LGTM , thanks!

**Pull request merged**